### DISCUSSION TITLE
Game recording and replay

### DISCUSSION THREAD

**Discussion by FreeLikeGNU** [bafolts/terraforming-mars#445] [Ideas]
It would be nice to replay a finished game and see turn by turn the moves.  The log is kind of general and fine for in-game, but a replay could provide an automated replication of the game to help squash bugs, make tutorials and help new players understand the game.   If there is already something like this in the game, forgive my ignorance.

**Comment by vincentneko** [comment]
I second this.
There is nothing like this at the moment.

Having this could also open possibilities for machine learning. I'm dreaming to have bots to play against.

**Comment by dabewi** [comment]
I would really really like this. I also share the same dream as vincentneko: Bot battles would be very nice.

**Comment by bafolts** [comment]
This project doesn't have much of a roadmap but I agree that heading towards a goal of bots to play against would be great. We have no bots right now. First introducing even a random bot that makes random moves may help with learning what changes need to be made to support bots. We have moved most things to use the interrupt system. Once everything were using the interrupt system this may help see the progression of a game as a queue.

This will require collecting these logs somewhere. The project is hosted on the heroku app right now and I don't think the hosting costs to run this project would be that demanding. I think @vincentneko hosts some game server somewhere.

**Comment by vincentneko** [comment]
I use Amazon AWS EC2 free tier, with docker. I'm currently testing this [Postgrest](http://postgrest.org/en/v6.0/), with the idea to store game results. I guess it could also be used for storing the logs.

**Comment by AbnRgr1978** [comment]
Non programmer chiming in.  At the couple of games ends I was able to get to I did see a game ID.  I was thinking this would be a good tool for tracking down bug issues.
Have not seen it any time other than game end which limits its use for reporting bugs, if that was an option.  Both 2 players games and several solo games I have tried ended in unrecoverable server issues.  Even with undo option selected we did not see an undo button that was mentioned elsewhere.

**Comment by bafolts** [comment]
If playing on the public terraforming-mars.herokuapp.com this should be more stable now. I have updated the database used and app level from the free plan which will make things more stable. With a provided game id during a bug report we should see the last game state that was saved in the database which should make recreating issues easier since they now shouldn't be related to the server restarting (unless we have a bug there).
